### DISCUSSION TITLE
MenuItem link for lens

### DISCUSSION THREAD

**Discussion by hivokas** [laravel/nova-issues#4280] [Ideas & Feature Requests]
It would be great to me able to pass lens class as a second param to `MenuItem::resource()` to add the menu link to resource lens.

**Comment by davidhemphill** [comment]
Agree, but it's a little tricky since lenses can be used on multiple resources (they are similar to filters in this regard). They are not explicitly tied to an individual resource.

**Reply by juse-less** [reply]
It would be nice if it was possible to create Resource-specific Lenses. Not sure where (if) one could make a feature request for that. 🙂
My use case for Lenses are pretty much always for specific Resources.
It's probably the wrong usage, but.. I create them to show 'reports' for specific data, or show data and Actions to specific tasks.
Technically speaking, it obviously wouldn't change much (nothing?), as they'd still be applied the same way, just throw an Exception (or so), when used on another Resource.

**Reply by hivokas** [reply]
Yeah, that's why I think this method should accept both resource class name and lens class name: https://github.com/laravel/nova-issues/discussions/4280#discussioncomment-2848430

**Comment by juse-less** [comment]
FWIW - I've created a method in my `NovaServiceProvider` (since I only need it there), to easily create Resource-specific Lens links.

```php
/**
 * @param  class-string<Lens>  $lens
 * @param  class-string<Resource>  $resource
 */
private static function resourceLensUrl(string $lens, string $resource): string
{
    return "/resources/{$resource::uriKey()}/lens/{$lens::make()->uriKey()}";
}
```

```php
// ...
MenuItem::link(__('Link Title'), path: self::resourceLensUrl(
    MyLens::class,
    MyResource::class,
));
```

Figuring out a more permanent solution to call directly on a `MenuItem` would obviously be preferred. 🙂

**Comment by hivokas** [comment]
Here is my solution:

```php
<?php

namespace App\Nova\Support\Menu;

use Laravel\Nova\Lenses\Lens;

class MenuItem extends \Laravel\Nova\Menu\MenuItem
{
    public static function lens($resourceClass, $lensClass): self
    {
        /** @var Lens $lens */
        $lens = new $lensClass;

        return self::resource($resourceClass)
            ->path('/resources/'.$resourceClass::uriKey().'/lens/'.$lens->uriKey())
            ->name($lens->name());
    }
}
```

**Comment by hivokas** [comment]
@davidhemphill thanks for implementing!

**Reply by davidhemphill** [reply]
You're welcome!
